modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py

```diff
@@ -137,6 +137,112 @@ def get_indices(cls, axis, partitions, index_func=None):
         new_idx = ray.get(new_idx)
         return new_idx[0].append(new_idx[1:]) if len(new_idx) else new_idx
 
+    @classmethod
+    def concat(cls, axis, left_parts, right_parts):
+        result = super(PandasOnRayDataframePartitionManager, cls).concat(
+            axis, left_parts, right_parts
+        )
+        return cls.rebalance_partitions(result)
+
+    @classmethod
+    def rebalance_partitions(cls, partitions):
+        from modin.config import NPartitions
+
+        heuristic = 1.5  # partitions can be 1.5x larger than ideal. Can be modified.
+        print(
+            f"partitions before rebalance: { [[obj._length_cache for obj in row] for row in partitions]}"
+        )
+        if partitions.shape[0] > NPartitions.get() * heuristic:
+            lengths = [[obj._length_cache for obj in row] for row in partitions]
+            if partitions.shape[0] % NPartitions.get() == 0:
+                ideal_partitions_per_axis_partition = (
+                    partitions.shape[0] // NPartitions.get()
+                )
+            else:
+                ideal_partitions_per_axis_partition = (
+                    partitions.shape[0] // NPartitions.get() + 1
+                )
+            # Naive rebalance
+            if any(l is None for row in lengths for l in row):
+                return np.array(
+                    [
+                        cls.column_partitions(
+                            partitions[i : i + ideal_partitions_per_axis_partition],
+                            full_axis=False,
+                        )
+                        for i in range(
+                            0, partitions.shape[0], ideal_partitions_per_axis_partition
+                        )
+                    ]
+                )
+            else:
```

Comment:
Can `def compute_schema(axis_lengths, axis):` function from the [PR](https://github.com/modin-project/modin/pull/3720) be reused here? It's tested and takes up less space.

Reply:
@anmyachev it's a great idea.